game-app/game-core/src/main/java/games/strategy/engine/ClientFileSystemHelper.java

```diff
@@ -38,19 +40,32 @@ public static Path getRootFolder() {
     }
   }
 
-  private static Path getCodeSourceFolder() throws IOException {
-    final ApplicationContext applicationContext = Services.loadAny(ApplicationContext.class);
-    final @Nullable CodeSource codeSource =
-        applicationContext.getMainClass().getProtectionDomain().getCodeSource();
-    if (codeSource == null) {
-      throw new IOException("code source is not available");
-    }
+  public static void setCodeSourceFolder(final Path sourceFolder) {
+    codeSourceLocation = sourceFolder;
+  }
 
-    final Path codeSourceLocation;
-    try {
-      codeSourceLocation = Path.of(codeSource.getLocation().toURI());
-    } catch (final URISyntaxException e) {
-      throw new IOException("code source location URI is malformed", e);
+  /**
+   * TripleA stores two folders, one for user content that survives between game installs, and a
```

Comment:
What would you like me to do?

Reply:
Is there a strong reason to use 'code source folder' over 'root folder'? If no, then using the existing method to get 'root folder' should be the same thing.

Reply:
Does your comment actually apply to `EngineImageLoader.loadImage` and you suggest calling `getRootFolder` instead of `getCodeSourceFolder`? 

If so: That would be possible, and leads to the following question: `getRootFolder` actually does not return the path to the root folder, but to a file within the root folder. When called by `ResourceLoader` the actual root folder is extracted from the return value of `getRootFolder` by `ResourceLoader.findDirectory`. When `getRootFolder` is called by `ProcessRunnerUtil` or `DecorationPlacer` this is not done, and I have no idea how it works in this context (or if it works at all).

To me the best solution seems to be to have `getRootFolder` actually return the root folder. But I don't know how to produce the calls from `ProcessRunnerUtil` and `DecorationPlacer`. So I couldn't test it thoroughly.

The IMHO second best solution would be to move `findDirectory` to `ClientFileSystemHelper `so `EngineImageLoader`  can call it without introducing a dependency to `ResourceLoader`.

What do you think?

Reply:
> and you suggest calling getRootFolder instead of getCodeSourceFolder?

Yes

> getRootFolder actually does not return the path to the root folder, but to a file within the root folder. 
> and I have no idea how it works in this context (or if it works at all).

You might have stumbled on a very major bug. The javadoc, the name of the method, it all indicates we are finding a folder. Further, the file was added (not by me, but I recall the background) to make finding the root folder deterministics and no longer guess work (as it really does vary depending on how the game is run and if you're running from source or from a jar or from the install4j installer).

> To me the best solution seems to be to have getRootFolder actually return the root folder. But I don't know how to produce the calls from ProcessRunnerUtil and DecorationPlacer

I agree

Reply:
Merry Christmas, all!

> You might have stumbled on a very major bug. The javadoc, the name of the method, it all indicates we are finding a folder.

This comment caught my attention because, indeed, `getRootFolder()` should be returning the path to the root folder, not the path to the _.triplea-root_ file within the root folder. From a very cursory search of the _ClientFileSystemHelper.java_ history, it appears this regression may have been introduced in #8862. It is there that the helper method `getFolderContainingFileWithName()` was replaced with `FileUtils#findFileInParentFolders()`. Simply from the method names, it appears the former returned a folder path while the latter returns a file path. Hope that helps.